conjure-undertow-processor/src/main/java/com/palantir/conjure/java/undertow/processor/ConjureUndertowAnnotationProcessor.java

```diff
@@ -0,0 +1,223 @@
+    }
+
+    @Override
+    public boolean process(Set<? extends TypeElement> _annotations, RoundEnvironment roundEnv) {
+        if (roundEnv.processingOver()) {
+            return false;
+        }
+
+        KeyedStream.of(roundEnv.getElementsAnnotatedWith(Handle.class))
+                .mapKeys(Element::getEnclosingElement)
+                .collectToSetMultimap()
+                .asMap()
+                .forEach((interfaceElement, annotatedMethods) -> {
+                    JavaFile javaFile;
+                    try {
+                        javaFile = generateUndertowServiceEndpoints(interfaceElement, annotatedMethods);
+                    } catch (Throwable e) {
+                        error("Code generation failed", interfaceElement, e);
+                        return;
+                    }
+
+                    try {
+                        Goethe.formatAndEmit(javaFile, filer);
+                    } catch (GoetheException e) {
+                        error("Failed to format generated code", interfaceElement, e);
+                    }
+                });
+
+        return false;
+    }
+
+    private JavaFile generateUndertowServiceEndpoints(
+            Element annotatedInterface, Collection<? extends Element> annotatedMethods) {
+        validationStep(ctx -> {
+            for (Element element : annotatedMethods) {
+                if (!element.getKind().equals(ElementKind.METHOD)) {
+                    ctx.reportError("Only methods can be annotated with @Handle", element);
+                } else if (element.getModifiers().contains(Modifier.PRIVATE)
+                        || element.getModifiers().contains(Modifier.PROTECTED)) {
+                    ctx.reportError(
+                            "Methods annotated with '@Handle' must be accessible to classes in the same "
+                                    + "package, neither 'private' nor 'protected' modifiers are allowed.",
+                            element);
+                } else if (element.getModifiers().contains(Modifier.STATIC)) {
+                    ctx.reportError("The '@Handle' annotation does not support static methods.", element);
+                }
+            }
+        });
+
+        List<EndpointDefinition> endpoints = processingStep(ctx -> {
+            EndpointDefinitions endpointDefinitions = new EndpointDefinitions(ctx, elements, types);
+            List<Optional<EndpointDefinition>> maybeEndpoints = annotatedMethods.stream()
+                    .map(MoreElements::asExecutable)
+                    .map(endpointDefinitions::tryParseEndpointDefinition)
+                    .collect(Collectors.toList());
+
+            Preconditions.checkArgument(
+                    maybeEndpoints.stream()
+                            .filter(Predicates.not(Optional::isPresent))
+                            .collect(Collectors.toList())
+                            .isEmpty(),
+                    "Failed validation");
```

Comment:
Maybe instead of returning an empty optional on failed validation, we should just throw inside `EndpointDefinitions#tryParseEndpointDefinition` where we have additional error context?